### PR TITLE
Ffi revision

### PR DESCRIPTION
**Pull request by abizjak**
## Purpose

Restructure wasm-chain-integration to prepare for upcoming contract state and view function changes.

## Changes

There are two main changes
- in ffi, module artifacts are no longer passed serialized, instead we give out pointers and the module is managed on the rust heap. This is necessary to support resumption of execution.
- put existing execution functions into a v0 module. Changes in v1 will be substantial and will mean we will have to rewrite most of the host functions.

This will have an accompanying pull request in the node.

## Checklist

- [x] My code follows the style of this project.
- [x] The code compiles without warnings.
- [x] I have performed a self-review of the changes.
- [x] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [x] (If necessary) I have updated the CHANGELOG.

### CODE REVIEW

File: wasm-chain-integration/src/v0/ffi.rs [hunk 1]
```diff
@@ -134,11 +146,11 @@ unsafe extern "C" fn call_receive(
...
 /// This function is safe provided all the supplied pointers are not null and
 /// the `wasm_bytes_ptr` points to an array of length at least `wasm_bytes_len`.
-unsafe extern "C" fn validate_and_process(
+unsafe extern "C" fn validate_and_process_v0(
     wasm_bytes_ptr: *const u8,
     wasm_bytes_len: size_t,
-    artifact_len: *mut size_t, // this is the length of the artifact
-    output_len: *mut size_t,   // this is the total length of the output, artifact + exports.
+    output_len: *mut size_t, // this is the total length of the output byte array
+    output_artifact: *mut *const ArtifactV0,
```
Comment (Bargsteen) [inline wasm-chain-integration/src/v0/ffi.rs:153]: The doc comment should be updated as well.

File: wasm-chain-integration/src/v0/mod.rs [hunk 2]
```diff
@@ -0,0 +1,982 @@
...
+            out.extend_from_slice(&(v.len() as u32).to_be_bytes());
+            out.extend_from_slice(v);
+        }
+        out
+    }
+}
+
+#[derive(Clone, Default)]
+/// The Default instance of this type constructs an empty list of outcomes.
+pub struct Outcome {
```
Comment (Bargsteen) [inline wasm-chain-integration/src/v0/mod.rs:55]: How come this and the host structs are here instead of in `types.rs`? 😊

**Review by Bargsteen — CHANGES_REQUESTED** [review]
Looks good in general. Just a few minor issues.

File: wasm-chain-integration/src/v0/mod.rs (hunk 2 above)
Comment (abizjak) [reply wasm-chain-integration/src/v0/mod.rs:55]: Because I expect that they will change in v1 with view functions and additional details.

Comment (abizjak) [reply wasm-chain-integration/src/v0/mod.rs:55]: We might revise later once we have an implementation of v1.

**Pull request merged**